File: JavaScript/robot-simulator.js
class InvalidInputError extends Error {
    constructor(message) {
        super();
        this.message = message || 'Invalid Input';
    }
}
exports.InvalidInputError = InvalidInputError;

function isDirection(value) {
    return value === 'north' || value === 'east' || value === 'south' || value === 'west';
}

class Robot {
    constructor() {
        this.coordinates = [0, 0];
        this.bearing = 'north';
    }
  
    place({ x, y, direction }) {
        this.coordinates = [x, y];
        if (!isDirection(direction)) {
            throw new InvalidInputError(`Invalid direction: ${direction}`);
        }
      
        this.bearing = direction;
    }
  
    turnRight() {
        switch (this.bearing) {
            case 'north':
                this.bearing = 'east';
                break;
            case 'east':
                this.bearing = 'south';
                break;
            case 'south':
                this.bearing = 'west';
                break;
            case 'west':
                this.bearing = 'north';
                break;
        }
    }
  
    turnLeft() {
        switch (this.bearing) {
            case 'north':
                this.bearing = 'west';
                break;
            case 'west':
                this.bearing = 'south';
                break;
            case 'south':
                this.bearing = 'east';
                break;
            case 'east':
                this.bearing = 'north';
                break;
        }
    }
  
    advance() {
        switch (this.bearing) {
            case 'north':
                this.coordinates[1]++;
                break;
            case 'south':
                this.coordinates[1]--;
                break;
            case 'east':
                this.coordinates[0]++;
                break;
            case 'west':
                this.coordinates[0]--;
                break;
        }
    }
    evaluate(instructions) {
        for (let i = 0; i < instructions.length; i++) {
            const command = instructions[i].toUpperCase();
            switch (command) {
                case 'R':
                    this.turnRight();
                    break;
                case 'L':
                    this.turnLeft();
                    break;
                case 'A':
                    this.advance();
                    break;
                default:
                    throw new InvalidInputError(`Invalid command: ${command}`);
            }
        }
    }
}
exports.Robot = Robot;

const r = new Robot();
r.place({ x: 7, y: 3, direction: 'north' });
r.evaluate("RAALAL");

console.log(r);

File: C#/Robot Simulator/RobotSimulator.cs
using System;

public enum Direction
{
    North,
    East,
    South,
    West
}

public class RobotSimulator
{
    public Direction Direction { get; private set; } = Direction.North;

    public int X { get; private set; } = 0;

    public int Y { get; private set; } = 0;


    public RobotSimulator() { }

    public RobotSimulator(Direction direction, int x, int y)
    {
        this.Direction = direction;
        this.X = x;
        this.Y = y;
    }

    public void TurnRight()
    {
        switch (this.Direction)
        {
            case Direction.North:
                this.Direction = Direction.East;
                break;
            case Direction.East:
                this.Direction = Direction.South;
                break;
            case Direction.South:
                this.Direction = Direction.West;
                break;
            case Direction.West:
                this.Direction = Direction.North;
                break;
        }
    }

    public void TurnLeft()
    {
        switch (this.Direction)
        {
            case Direction.North:
                this.Direction = Direction.West;
                break;
            case Direction.West:
                this.Direction = Direction.South;
                break;
            case Direction.South:
                this.Direction = Direction.East;
                break;
            case Direction.East:
                this.Direction = Direction.North;
                break;
        }
    }

    public void Advance()
    {
        switch (this.Direction)
        {
            case Direction.North:
                this.Y++;
                break;
            case Direction.South:
                this.Y--;
                break;
            case Direction.East:
                this.X++;
                break;
            case Direction.West:
                this.X--;
                break;
        }
    }

    public void Move(string instructions)
    {
        foreach(var instruction in instructions)
        {
            var command = Char.ToUpper(instruction);

            switch (command)
            {
                case 'R':
                    this.TurnRight();
                    break;
                case 'L':
                    this.TurnLeft();
                    break;
                case 'A':
                    this.Advance();
                    break;
                default:
                    throw new ArgumentException("Invalid instruction in string.");
            }
        }
    }
}